community/kernel/src/main/java/org/neo4j/kernel/impl/api/KernelTransactionImplementation.java

```diff
@@ -312,7 +316,7 @@ public void markForTermination()
             if ( stillSameTransaction && canBeTerminated() )
             {
                 failure = true;
-                terminated = true;
+                terminationReason = reason;
```

Comment:
If we do that, then I think all loads and stores with the `terminationReason` field will be protected by the lock, so the field won't need to be volatile.